toppra/constraint/cartesian_constraints.py

```diff
@@ -0,0 +1,43 @@
+import numpy as np
+from .linear_second_order import SecondOrderConstraint
+
+class CartesianSpeedConstraint(SecondOrderConstraint):
+    """
+    This class implements a constraint on the magnitudes of the linear & angular
+    Cartesian velocity vectors of one of the robot's parts (link, joint, etc).
+    
+    The forward kinematic velocity is be provided via a callback, which makes
+    this constraint agnostic of the robot's geometry and agnostic of whatever
+    FK algorithm is used.
+    """
+
+    def __init__(self, fk, linear_speed_max, angular_speed_max, dof):
+        """Initialize the constraint.
+
+        Parameters
+        ----------
+        fk: (np.ndarray, np.ndarray) -> (float, float)
+            The "FK" function that receives joint positions and velocities as
+            inputs and outputs the magnitude of the linear and angular
+            velocity vectors for some the monitored part of the robot.
+        linear_speed_max: float
+            The max linear speed allowed for the monitored part.
+        angular_speed_max: float
+            The max angular speed allowed for the monitored part.
+        dof: int
+            The dimension of the joint position.
+        """
+        super(CartesianSpeedConstraint, self).__init__(self.invdyn, self.constraintf, self.constraintg, dof)
+        self.fk = fk
+        self.linear_speed_max = linear_speed_max
+        self.angular_speed_max = angular_speed_max
+
+    def invdyn(self, q, dq, ddq):
+        linear_speed, angular_speed = self.fk(q, dq)
+        return np.array([linear_speed**2, angular_speed**2])
```

Comment:
This seems strange. We shouldn't have to square the speeds nor the speed limits.

Reply:
I thought it was weird too. If I don't square then it doesn't seem to work at all. The below screenshots should both have a speed limits of 1m/s and 1rad/s.

![Without squaring](https://user-images.githubusercontent.com/6147146/99145030-0c9a7900-2663-11eb-933c-f894be3f7f70.png)

![With squaring](https://user-images.githubusercontent.com/6147146/99145047-2f2c9200-2663-11eb-8dbd-153ac549676b.png)

Reply:
I think you have to square it for the reason I mentioned in #101. Following the derivation of #146, you should be able to understand why.

Reply:
Thanks @jmirabel, I think I've got my head around it now